allow libbacktrace to be used when cross compiling the runtime

mherkazandjian opened the pull request:
This PR fixes issue #7916

to build for aarch64 the cmake variable  ``-DCMAKE_HOST=aarch64-linux-gnu`` should be passed that
is in turn propagated to the autotools (configure step)  for libbacktrace.

````bash
cmake .. \
    -DCMAKE_SYSTEM_NAME=Linux \
    -DCMAKE_SYSTEM_VERSION=1 \
    -DCMAKE_C_COMPILER=/usr/bin/aarch64-linux-gnu-gcc-8 \
    -DCMAKE_CXX_COMPILER=/usr/bin/aarch64-linux-gnu-g++-8 \
    -DCMAKE_FIND_ROOT_PATH=/usr/aarch64-linux-gnu \
    -DCMAKE_FIND_ROOT_PATH_MODE_PROGRAM=NEVER \
    -DCMAKE_FIND_ROOT_PATH_MODE_LIBRARY=ONLY \
    -DCMAKE_HOST=aarch64-linux-gnu  
````

I tested this to build the runtime for ``aarch64``  and ``riscv64``, the runtime ``.so`` was produced in both
cases with no errors displayed during the build.

tqchen commented:
@areusch @tkonolige @leandron please help to review this PR

junrushao1994 commented:
CC @tkonolige @xiebaiyuan The issue is possibly related to #7705.

leandron commented on cmake/libs/Libbacktrace.cmake:
```diff
@@ -21,7 +21,16 @@ ExternalProject_Add(project_libbacktrace
+                    "CC=${CMAKE_C_COMPILER}"
+                    "CFLAGS=${CMAKE_C_FLAGS}"
+                    "LDFLAGS=${CMAKE_EXE_LINKER_FLAGS}"
+                    "CPP=${CMAKE_C_COMPILER} -E"
```
Maybe this should actually point to the CXX version?

```suggestion
                    "CPP=${CMAKE_CXX_COMPILER} -E"
```

mherkazandjian replied:
Hi @leandron 

I think you are right, i'll do the changes and also update the documention.
On another note, i think an end to end example on inference on a remote machine would be great. 
i could probably contribute that end to end example if it does not exist, does it?

mherkazandjian replied:
i looked into the github repo of  [libbacktrace](https://github.com/ianlancetaylor/libbacktrace) and it seems here ``CPP``
refers to the C preprocess (not C++), so it think it is correct to keep it as is.

leandron reviewed (commented):
Thanks for this PR. As the default value for libbacktrace is `AUTO`, this patch enables cross-compilation of the source in the original cloning state - great!

On a broader view, it would be awesome to also update the documentation such as https://tvm.apache.org/docs/install/from_source.html, with some guidance for cross-compilation as there are non-obvious steps to be taken, such as the need for `-DCMAKE_HOST=...` to be populated, for example.

What do you think?

leandron commented:
> On another note, i think an end to end example on inference on a remote machine would be great.
i could probably contribute that end to end example if it does not exist, does it?

Yes, that would be really useful, specially if it uses TVMC rather than low level APIs, IMHO.

leandron commented on cmake/libs/Libbacktrace.cmake:
```diff
@@ -21,7 +21,16 @@ ExternalProject_Add(project_libbacktrace
+                    "CPP=${CMAKE_C_COMPILER} -E"
+                    "NM=${CMAKE_NM}"
+                    "STRIP=${CMAKE_STRIP}"
+                    "CPPFLAGS=${CMAKE_C_FLAGS}"
```
Just noticed this one probably needs updating as well.

mherkazandjian replied:
> Just noticed this one probably needs updating as well.

done

tkonolige reviewed (approved):
These changes look good to me.

tkonolige commented:
btw, I tried these same changes when debugging #7705 but they did not fix it.

areusch commented on cmake/libs/Libbacktrace.cmake:
```diff
@@ -21,7 +21,16 @@ ExternalProject_Add(project_libbacktrace
+                    "NM=${CMAKE_NM}"
+                    "STRIP=${CMAKE_STRIP}"
+                    "CPPFLAGS=${CMAKE_C_FLAGS}"
+                    "--host=${MACHINE_NAME}"
```
I don't see this one documented in cmake, do you need to supply it with `cmake -DMACHINE_NAME=`?

mherkazandjian replied:
yes, it should be specified when executing cmake

![image](https://user-images.githubusercontent.com/3264637/116225445-1ff15a80-a752-11eb-9ae5-fac0eac7236b.png)

mherkazandjian replied:
actually i am not sure is ``-DCMAKE_HOST`` is neccessary, but it does not hurt (i guess)

areusch replied:
okay cool. would it be possible for you to add the steps you took to cross-compile to the [install from source](https://tvm.apache.org/docs/install/from_source.html) docs?

mherkazandjian replied:
yes, working on it, i will push in a few sec.

mherkazandjian replied:
done, i built the doc locally and it is rendered ok

areusch reviewed (commented):
thanks @mherkazandjian for the PR! just one question, I agree with @leandron that it would be great to provide some docs.

leandron commented on docs/deploy/index.rst:
```diff
@@ -46,20 +54,105 @@ on a Linux based embedded system such as Raspberry Pi:
+
+.. code-block:: bash
+
+   file file libtvm_runtime.so
```
Is `file file` here intentional?

mherkazandjian replied:
fixed

leandron commented on docs/deploy/index.rst:
```diff
@@ -46,20 +54,105 @@ on a Linux based embedded system such as Raspberry Pi:
+-----------------------------------------------------
+
+In the example :ref:`above <build-tvm-runtime-on-target-device>` the runtime library was 
+compiled on a raspberry pi. Producing the runtime library can be done much faster on 
```
```suggestion
compiled on a Raspberry Pi. Producing the runtime library can be done much faster on 
```

mherkazandjian replied:
done, tnx

leandron commented on docs/deploy/index.rst:
```diff
@@ -46,20 +54,105 @@ on a Linux based embedded system such as Raspberry Pi:
+hosts that have beefy processors (such as laptops, workstation) instead of target
+devices such as a raspberry pi. In-order to cross compile the runtime the toolchain
+for the target device must be installed. After installing the correct toolchain
+the main difference compared to compiling natively is to pass a bunch of command
```
```suggestion
the main difference compared to compiling natively is to pass some additional command
```

mherkazandjian replied:
done

leandron commented on docs/deploy/index.rst:
```diff
@@ -46,20 +54,105 @@ on a Linux based embedded system such as Raspberry Pi:
+devices such as a raspberry pi. In-order to cross compile the runtime the toolchain
+for the target device must be installed. After installing the correct toolchain
+the main difference compared to compiling natively is to pass a bunch of command
+line argument to cmake that specify the compiler(s) to be used. For reference
```
```suggestion
line argument to cmake that specify a toolchain to be used. For reference,
```

mherkazandjian replied:
done

leandron commented on docs/deploy/index.rst:
```diff
@@ -46,20 +54,105 @@ on a Linux based embedded system such as Raspberry Pi:
+the main difference compared to compiling natively is to pass a bunch of command
+line argument to cmake that specify the compiler(s) to be used. For reference
+building the TVM runtime library on a modern laptop (using 8 threads) for ``aarch64``
+takes around 20 seconds vs ~10 min to build the runtime on a raspberry pi 4.
```
```suggestion
takes around 20 seconds vs ~10 min to build the runtime on a Raspberry Pi 4.
```

leandron commented on docs/deploy/index.rst:
```diff
@@ -46,20 +54,105 @@ on a Linux based embedded system such as Raspberry Pi:
+In the example :ref:`above <build-tvm-runtime-on-target-device>` the runtime library was 
+compiled on a raspberry pi. Producing the runtime library can be done much faster on 
+hosts that have beefy processors (such as laptops, workstation) instead of target
+devices such as a raspberry pi. In-order to cross compile the runtime the toolchain
```
```suggestion
devices such as a Raspberry Pi. In-order to cross compile the runtime, the toolchain
```

mherkazandjian replied:
done

leandron commented on docs/deploy/index.rst:
```diff
@@ -46,20 +54,105 @@ on a Linux based embedded system such as Raspberry Pi:
+compiled on a raspberry pi. Producing the runtime library can be done much faster on 
+hosts that have beefy processors (such as laptops, workstation) instead of target
+devices such as a raspberry pi. In-order to cross compile the runtime the toolchain
+for the target device must be installed. After installing the correct toolchain
```
```suggestion
for the target device must be installed. After installing the correct toolchain,
```

mherkazandjian replied:
done

leandron commented on docs/install/from_source.rst:
```diff
@@ -51,27 +51,31 @@ Build the Shared Library
 
 We use cmake to build the library.
-The configuration of TVM can be modified by `config.cmake`.
+The configuration of TVM can be modified by `config.cmake` (and/or by passing cmake flags to the command line):
```
```suggestion
The configuration of TVM can be modified by `config.cmake` and/or by passing cmake flags to the command line:
```

mherkazandjian replied:
done

leandron reviewed (commented):
A few suggestions on the tutorial wording, feel free to decide whether you want to use it. @hogepodge recently did some work on tutorials and might be interested as well.

Apart from the suggestions, LGTM.

mherkazandjian commented:
@leandron tnx for the feedback. actually the CI build is failing, i think the in the CI we (I) should modify the cmake flags so that
libbacktrace now builds (it is failing in the CI). I will do it now.

leandron reviewed (commented):
I happy to approve this PR. Did you check @tkonolige's comment about building with debug flags?

mherkazandjian commented:
> I happy to approve this PR. Did you check @tkonolige's comment about building with debug flags?

great. thank you.
i looked into the comment of @tkonolige but i thought it is not strictly related to this issue in the sense that 
this PR would not affect that issue negatively. I will look into it closely.

i have a question though, in the CI it looks like the arm build is failing
https://ci.tlcpack.ai/blue/organizations/jenkins/tvm/detail/PR-7917/7/pipeline/#step-163-log-112
should i try to fix it? please advise.

tkonolige commented:
@mherkazandjian I'm sorry it took me so long to remember this, but I don't think this PR will work on macOS. Here is what I had in my version of this:

```
# On MacOS, the default C compiler (/usr/bin/cc) is actually a small script that dispatches to a
# compiler the default SDK (usually /Library/Developer/CommandLineTools/usr/bin/ or
# /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/). CMake
# automatically detects what is being dispatched and uses it instead along with all the flags it
# needs. CMake makes this second compiler avaliable through the CMAKE_C_COMPILER variable, but it
# does not make the necessary flags available. This leads to configuration errors in libbacktrace
# because it can't find system libraries. Our solution is to detect if CMAKE_C_COMPILER lives in
# /Library or /Applications and switch to the default compiler instead.
if(CMAKE_SYSTEM_NAME MATCHES "Darwin" AND (CMAKE_C_COMPILER MATCHES "^/Library"
  OR CMAKE_C_COMPILER MATCHES "^/Applications"))
  set(c_compiler "/usr/bin/cc")
else()
  set(c_compiler "${CMAKE_C_COMPILER}")
endif()
```

Note that this solution is not quite correct. You probably want to set the sdk path to whatever the user set it to.

mherkazandjian commented:
> btw, I tried these same changes when debugging #7705 but they did not fix it.

hi @tkonolige 


> @mherkazandjian I'm sorry it took me so long to remember this, but I don't think this PR will work on macOS. Here is what I had in my version of this:
> 
> ```
> # On MacOS, the default C compiler (/usr/bin/cc) is actually a small script that dispatches to a
> # compiler the default SDK (usually /Library/Developer/CommandLineTools/usr/bin/ or
> # /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/). CMake
> # automatically detects what is being dispatched and uses it instead along with all the flags it
> # needs. CMake makes this second compiler avaliable through the CMAKE_C_COMPILER variable, but it
> # does not make the necessary flags available. This leads to configuration errors in libbacktrace
> # because it can't find system libraries. Our solution is to detect if CMAKE_C_COMPILER lives in
> # /Library or /Applications and switch to the default compiler instead.
> if(CMAKE_SYSTEM_NAME MATCHES "Darwin" AND (CMAKE_C_COMPILER MATCHES "^/Library"
>   OR CMAKE_C_COMPILER MATCHES "^/Applications"))
>   set(c_compiler "/usr/bin/cc")
> else()
>   set(c_compiler "${CMAKE_C_COMPILER}")
> endif()
> ```
> 
> Note that this solution is not quite correct. You probably want to set the sdk path to whatever the user set it to.

ok tnx for the update.
Actually the CI is failing for all the builds (see below) not just for arm as i was thinking.
![image](https://user-images.githubusercontent.com/3264637/116439353-bc4d5700-a84f-11eb-801e-bdfbe7ef34f1.png)
i think we need to pass ``-DMACHINE_NAME`` to cmake in the CI build scripts too.

tkonolige commented:
@mherkazandjian We should not require `MACHINE_NAME` be passed in as part of the build. The cmake script should check if it is set and then add the correct configure flags.

mherkazandjian commented:
> @mherkazandjian We should not require `MACHINE_NAME` be passed in as part of the build. The cmake script should check if it is set and then add the correct configure flags.

@tkonolige i fixed the issue with the CI, the ``CPPFLAGS=${CMAKE_CXX_FLAGS}`` was the cause, I removed it since
``g++`` is not used/needed for libbacktrace (it is written in C).

the macOS build seems to be passing 
![image](https://user-images.githubusercontent.com/3264637/116488928-eaed2100-a893-11eb-9f8c-725439b967df.png)

following up on your comment https://github.com/apache/tvm/pull/7917#issuecomment-828562180 
you mean including the snippet below in ``Libbacktrace.cmake`` (although ) fixes the issue for macOS? what did you mean
by "although not correct?" 
i do not have access to a mac.

````cmake
if(CMAKE_SYSTEM_NAME MATCHES "Darwin" AND (CMAKE_C_COMPILER MATCHES "^/Library"
  OR CMAKE_C_COMPILER MATCHES "^/Applications"))
  set(c_compiler "/usr/bin/cc")
else()
  set(c_compiler "${CMAKE_C_COMPILER}")
endif()
````

mherkazandjian commented:
@tqchen it looks like the docker daemon did not start in the CI instance
https://ci.tlcpack.ai/blue/organizations/jenkins/tvm/detail/PR-7917/9/pipeline/259
how can one re-trigger the build? it looks like an issue with the CI machine/vm itself.

areusch commented:
@mherkazandjian i think the CI is having some trouble now, feel free to push an empty commit to retriever

mherkazandjian commented:
> @mherkazandjian i think the CI is having some trouble now, feel free to push an empty commit to retriever

cool. i did a dummy commit to trigger the build, the CI checks pass. I am not sure @tkonolige is ok with merging but 
i don't see any macOS issues in the CI (maybe i am missing something). Please feel free to merge whenever you want.

hogepodge commented on docs/deploy/index.rst:
```diff
@@ -46,20 +54,105 @@ on a Linux based embedded system such as Raspberry Pi:
-Note that we type `make runtime` to only build the runtime library.
+Note that we type ``make runtime`` to only build the runtime library.
+
+It is also possible to cross compile the runtime. Cross compiling
```
It would be helpful to describe why you would want to cross compile the runtime.

mherkazandjian replied:
i added a few lines on that and also i added a diagram the visually describes the process, i think that 
improved the clarify compared to describing it by words

this is how that is rendered in the documentation

![image](https://user-images.githubusercontent.com/3264637/116707916-7b7c4c00-a9cf-11eb-902f-c8aee573ee99.png)

please let me know if it is ok to include the figure in ``deploy/images`` (i added that folder) in some other places figures
are directly included from github rather than from files stores in the repo.

mherkazandjian replied:
ok, if you are ok with the image i can send a PR there.

mherkazandjian replied:
https://github.com/tlc-pack/web-data/pull/5

mherkazandjian replied:
thank you @tqchen  for merging https://github.com/tlc-pack/web-data/pull/5 
i updated the url, the doc build looks good locally.

hogepodge commented on docs/deploy/index.rst:
```diff
@@ -46,20 +54,105 @@ on a Linux based embedded system such as Raspberry Pi:
+
+In the example :ref:`above <build-tvm-runtime-on-target-device>` the runtime library was 
+compiled on a Raspberry Pi. Producing the runtime library can be done much faster on 
+hosts that have beefy processors (such as laptops, workstation) instead of target
```
I would avoid using jargon like "beefy" and prefer phrases like "higher performance".

hogepodge commented:
My apologies for the late docs review. Thanks for pinging me on it.

mherkazandjian commented:
> My apologies for the late docs review. Thanks for pinging me on it.

no worries. tnx for the feedback. i'll check your feedback in a bit.

tkonolige commented:
@mherkazandjian That snippet will fix the issue. `/usr/bin/cc` looks at the environment variable `SDKROOT` and uses the `cc` in that directory. So if the user switches their `SDKROOT` between configuring and building tvm, then they will have an issue.

mherkazandjian commented:
> @mherkazandjian That snippet will fix the issue. `/usr/bin/cc` looks at the environment variable `SDKROOT` and uses the `cc` in that directory. So if the user switches their `SDKROOT` between configuring and building tvm, then they will have an issue.

i added the snippet to the cmake config of libbacktrace.

hogepodge commented:
We put images into the https://github.com/tlc-pack/web-data repository.

mherkazandjian commented:
@hogepodge did you get a chance to look at the changes? i think this PR is good to go (the CI tests pass too)

tkonolige commented on cmake/libs/Libbacktrace.cmake:
```diff
@@ -16,12 +16,28 @@
-                    "--prefix=${CMAKE_CURRENT_BINARY_DIR}/libbacktrace" --with-pic
+                    "--prefix=${CMAKE_CURRENT_BINARY_DIR}/libbacktrace"
+                    --with-pic
+                    "CC=${CMAKE_C_COMPILER}"
```
```suggestion
                    "CC=${c_compiler}"
```
You need this change to match the lines above.

tkonolige commented on cmake/libs/Libbacktrace.cmake:
```diff
@@ -16,12 +16,28 @@
+                    "CC=${CMAKE_C_COMPILER}"
+                    "CFLAGS=${CMAKE_C_FLAGS}"
+                    "LDFLAGS=${CMAKE_EXE_LINKER_FLAGS}"
+                    "CPP=${CMAKE_C_COMPILER} -E"
```
```suggestion
                    "CPP=${c_compiler} -E"
```

tkonolige commented on cmake/libs/Libbacktrace.cmake:
```diff
@@ -16,12 +16,28 @@
 # under the License.
 include(ExternalProject)
 
+
```
Could you add this comment:

```
# On MacOS, the default C compiler (/usr/bin/cc) is actually a small script that dispatches to a
# compiler the default SDK (usually /Library/Developer/CommandLineTools/usr/bin/ or
# /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/). CMake
# automatically detects what is being dispatched and uses it instead along with all the flags it
# needs. CMake makes this second compiler avaliable through the CMAKE_C_COMPILER variable, but it
# does not make the necessary flags available. This leads to configuration errors in libbacktrace
# because it can't find system libraries. Our solution is to detect if CMAKE_C_COMPILER lives in
# /Library or /Applications and switch to the default compiler instead.
```

mherkazandjian replied:
done, tnx.

tkonolige reviewed (changes requested):
You need a couple of changes to make the macOS workaround work.

hogepodge commented on docs/deploy/index.rst:
```diff
@@ -46,20 +54,118 @@ on a Linux based embedded system such as Raspberry Pi:
 After you get the TVM runtime library, you can link the compiled library
 
+Cross compiling a TVM model should not be confused with cross compiling the runtime
+shared library. 
```
Duplicated wording

tqchen commented:
gentle ping @mherkazandjian please follow up on the review comments and let us merge soon !

mherkazandjian commented:
> gentle ping @mherkazandjian please follow up on the review comments and let us merge soon !

Hi @tqchen 
sorry got distracted with some other work. I'll wrap this up before the end of the week.

mherkazandjian commented:
@tqchen i finished all the requested changes :)

tqchen commented:
@areusch @leandron @tkonolige @hogepodge please take another look and https://tvm.apache.org/docs/contribute/code_review.html?highlight=approve#approve-and-request-changes-explicitly

hogepodge commented on cmake/libs/Libbacktrace.cmake:
```diff
@@ -14,14 +14,39 @@
 # under the License.
+
+# On MacOS, the default C compiler (/usr/bin/cc) is actually a small script that dispatches to a
+# compiler the default SDK (usually /Library/Developer/CommandLineTools/usr/bin/ or
```
On MacOS, the default C compiler (/usr/bin/cc) is script that dispatches to the compiler
provided by the default SDK (usually /Library/Developer/CommandLineTools/usr/bin/ or

hogepodge commented on cmake/libs/Libbacktrace.cmake:
```diff
@@ -14,14 +14,39 @@
+# needs. CMake makes this second compiler avaliable through the CMAKE_C_COMPILER variable, but it
+# does not make the necessary flags available. This leads to configuration errors in libbacktrace
+# because it can't find system libraries. Our solution is to detect if CMAKE_C_COMPILER lives in
+# /Library or /Applications and switch to the default compiler instead.
```
Our solution is to detect if CMAKE_C_COMPILER is installed in

hogepodge commented on docs/deploy/index.rst:
```diff
@@ -25,12 +25,20 @@ as well as how to integrate it with your project.
 
-In order to integrate the compiled module, we **do not** need to build entire TVM on the target device. You only need to build the TVM compiler stack on your desktop and use that to cross-compile modules that are deployed on the target device.
+In order to integrate the compiled module, we **do not** need to build entire
+TVM on the target device. You only need to build the TVM compiler stack on your 
```
we **do not** need to build the entire TVM project on the target device. Rather, we only need to build the TVM compiler stack on a host device and use that to cross-compile modules that can be deployed to a target device.

hogepodge reviewed (changes requested):
Throughout this document, there's a mix of using "we" and "you" to describe the process of building TVM. This should be reconciled, preferably to use the 3rd person. That may be a larger task for another PR. A number of other comments to help clarify language and usage.

tqchen commented:
Thank you @mherkazandjian ! Thanks @tkonolige @hogepodge @areusch @leandron for reviewing
